snapcraft/plugins/ament.py

```diff
@@ -0,0 +1,197 @@
+
+    def pull(self):
+        """Pull the provided source, as well as the ROS2 underlay."""
+
+        super().pull()
+
+        # Pull ros2 underlay
+        self._bootstrapper.pull()
+
+    def clean_pull(self):
+        """Clean the fetched source and ROS2 underlay."""
+
+        super().clean_pull()
+
+        self._bootstrapper.clean()
+
+    def build(self):
+        """Bootstrap the ROS2 underlay, then use it to build the source.
+
+        This step can take a while.
+        """
+
+        super().build()
+
+        self._prepare_build()
+
+        self.run([
+            'ament', 'build', self.sourcedir, '--build-space', self.builddir,
+            '--install-space', self.installdir, '--cmake-args',
+            '-DCMAKE_BUILD_TYPE=Release'], env=self._build_env())
+
+        self._finish_build()
+
+    def _prepare_build(self):
+        # Bootstrap ros2, and migrate it into our installdir
+        ros2_install_dir = self._bootstrapper.build()
+        snapcraft.file_utils.link_or_copy_tree(
+            ros2_install_dir, self.installdir)
+
+        # Also rewrite the prefixes to point to the part installdir rather
+        # than the bootstrap area. Otherwise Ament embeds absolute paths which
+        # makes the resulting code non-relocatable.
+        snapcraft.file_utils.replace_in_file(
+            self.installdir, re.compile(r''),
+            re.compile(r'\${AMENT_CURRENT_PREFIX:=.*}'),
+            '${{AMENT_CURRENT_PREFIX:={}}}'.format(self.installdir))
+
+    def _finish_build(self):
+        # Set the AMENT_CURRENT_PREFIX throughout to a sensible default,
+        # removing part-specific directories that will clash with other parts.
+        snapcraft.file_utils.replace_in_file(
+            self.installdir, re.compile(r''),
+            re.compile(r'\${AMENT_CURRENT_PREFIX:=.*}'),
+            '${AMENT_CURRENT_PREFIX:=$SNAP}')
+
+    def env(self, root):
+        """Runtime environment for ROS2 apps."""
+
+        env = []
+
+        env.append('PYTHONUSERBASE="{}"'.format(root))
+        env.append('PYTHONPATH="{}:{}${{PYTHONPATH:+:$PYTHONPATH}}"'.format(
+            os.path.join(root, 'usr', 'lib', 'python3', 'dist-packages'),
+            os.path.join(root, 'lib', 'python3', 'dist-packages')))
+
+        # Each of these lines is prepended with an `export` when the
+        # environment is actually generated. In order to inject real shell code
+        # we have to hack it in by appending it on the end of an item already
+        # in the environment. FIXME: There should be a better way to do this.
+        env[-1] = env[-1] + '\n\n' + self._source_setup_sh(root)
+
+        return env
+
+    def _build_env(self):
+        env = os.environ.copy()
+        env['PYTHONPATH'] = os.path.join(
+            os.path.sep, 'usr', 'lib', 'python3', 'dist-packages')
+        return env
+
+    def _source_setup_sh(self, root):
+        return textwrap.dedent('''
+            if [ -f {ros_setup} ]; then
```

Comment:
What's the reason for the `-f` here? Can't you assume the setup.sh is already in place?

Reply:
Not necessarily. Remember that `env()` is used both to generate the wrappers in prime _as well_ as when you call `self.run()` to build. While setup.sh will definitely be there by the end, it has to _get_ there first.

Reply:
Very true. I had a feeling I was missing something. Thanks for clarifying.